src/Connection.ts

```diff
@@ -1,1488 +1,1884 @@
+
+		this._socket.on("log", (message) => {
+			this.props.onLog?.(message);
+			this.onLogHandlers.forEach((cb) => cb(message));
+		});
+
+		this._socket.on("error", (err: any) => {
+			let _err: string;
+
+			if (err == undefined) {
+				_err = "";
+			} else if (typeof err.toString !== "function") {
+				_err = err.toString();
+			} else {
+				_err = JSON.stringify(err);
+				console.error(`Received strange error: ${_err}`);
+			}
+
+			if (_err.includes("User not authorized")) {
+				this.authenticate();
+			} else {
+				window.alert(`Socket Error: ${err}`);
+			}
+		});
+
+		this._socket.on("connect_error", (err: any) =>
+			console.error(`Connect error: ${err}`),
+		);
+
+		this._socket.on("permissionError", (err) =>
+			this.onError({
+				message: "no permission",
+				operation: err.operation,
+				type: err.type,
+				id: err.id || "",
+			}),
+		);
+
+		this._socket.on("objectChange", (id, obj) => {
+			setTimeout(() => this.objectChange(id, obj), 0);
+		});
+
+		this._socket.on("stateChange", (id, state) => {
+			setTimeout(() => this.stateChange(id, state), 0);
+		});
+
+		this._socket.on("cmdStdout", (id, text) => {
+			this.onCmdStdoutHandler?.(id, text);
+		});
+
+		this._socket.on("cmdStderr", (id, text) => {
+			this.onCmdStderrHandler?.(id, text);
+		});
+
+		this._socket.on("cmdExit", (id, exitCode) => {
+			this.onCmdExitHandler?.(id, exitCode);
+		});
+	}
+
+	/**
+	 * Called internally.
+	 * @param isOk
+	 * @param isSecure
+	 */
+	private onPreConnect(isOk: boolean, isSecure: boolean) {
+		if (this._authTimer) {
+			clearTimeout(this._authTimer);
+			this._authTimer = null;
+		}
+
+		this.connected = true;
+		this.isSecure = isSecure;
+
+		if (this.waitForRestart) {
+			window.location.reload();
+		} else {
+			if (this.firstConnect) {
+				this.loadData();
+			} else {
+				this.props.onProgress?.(PROGRESS.READY);
+			}
+
+			this._subscribe(true);
+			this.onConnectionHandlers.forEach((cb) => cb(true));
+		}
+
+		this._waitForFirstConnectionPromise.resolve();
+	}
+
+	/**
+	 * Checks if the socket is connected.
+	 * @returns {boolean} true if connected.
+	 */
+	isConnected(): boolean {
+		return this.connected;
+	}
+
+	/**
+	 * Checks if the socket is connected.
+	 * @returns {Promise<void>} Promise resolves if once connected.
+	 */
+	waitForFirstConnection(): Promise<void> {
+		return this._waitForFirstConnectionPromise;
+	}
+
+	/**
+	 * Called internally.
+	 */
+	private async getUserPermissions(): Promise<any> {
+		return this.request({
+			// TODO: check if this should time out
+			commandTimeout: false,
+			executor: (resolve, reject) => {
+				this._socket.emit("getUserPermissions", (err, acl) => {
+					if (err) reject(err);
+					resolve(acl!);
+				});
+			},
+		});
+	}
+
+	/** Loads the important data and retries a couple of times if it takes too long */
+	private async loadData(): Promise<void> {
+		if (this.loaded) return;
+		const maxAttempts = 10;
+		for (let i = 1; i <= maxAttempts; i++) {
```

Comment:
This is a really bad algorithm, but we don't have to fix it now.

**But,** we could at least use `Promise.any(this.doLoadData(), wait(1000))` so we don't always wait 1 second - just for fun...

Reply:
And for the future: in `doLoadData()` we should check if some data is already loaded and then skip loading it again - otherwise we simply create even more traffic.

Reply:
I'll switch the return and the wait for now :)
This algorithm is basically the old code but in an async loop instead of some callback madness.

Reply:
Oh, you're right - it actually doesn't matter how long `loadData()` takes as it is never awaited 🤦‍♂️